core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java

```diff
@@ -0,0 +1,267 @@
+    public static final String SMART_TESTING_APPLY_TO = "smart.testing.apply.to";
+    public static final String SMART_TESTING_VERSION = "smart.testing.version";
+    public static final String SMART_TESTING_DISABLE = "smart.testing.disable";
+    public static final String SMART_TESTING_DEBUG = "smart.testing.debug";
+
+    public static final String SMART_TESTING_YML = "smart-testing.yml";
+    public static final String SMART_TESTING_YAML = "smart-testing.yaml";
+
+    private String[] strategies = new String[0];
+    private RunMode mode;
+    private String applyTo;
+
+    private Boolean disable;
+    private Boolean debug;
+
+    private Report report;
+    private Scm scm;
+
+    public String[] getStrategies() {
+        return strategies;
+    }
+
+    public void setStrategies(String... strategies) {
+        this.strategies = strategies;
+    }
+
+    public RunMode getMode() {
+        return mode;
+    }
+
+    public void setMode(RunMode mode) {
+        this.mode = mode;
+    }
+
+    public String getApplyTo() {
+        return applyTo;
+    }
+
+    public void setApplyTo(String applyTo) {
+        this.applyTo = applyTo;
+    }
+
+    public Boolean isDisable() {
+        return disable != null && disable;
+    }
+
+    public void setDisable(Boolean disable) {
+        this.disable = disable;
+    }
+
+    public Boolean isDebug() {
+        return debug != null && debug;
+    }
+
+    public void setDebug(Boolean debug) {
+        this.debug = debug;
+    }
+
+    public Report getReport() {
+        return report;
+    }
+
+    public void setReport(Report report) {
+        this.report = report;
+    }
+
+    public Scm getScm() {
+        return scm;
+    }
+
+    public void setScm(Scm scm) {
+        this.scm = scm;
+    }
+
+    static Configuration fromSystemProperties() {
+        final Configuration configuration = new Configuration();
+
+        if (System.getProperty(SMART_TESTING) != null) {
+            final String strategies = System.getProperty(SMART_TESTING).toLowerCase();
+
+            if (containsAnyStrategy(strategies)) {
+                configuration.strategies = strategies.split("\\s*,\\s*");
+            }
+        }
+
+        if (System.getProperty(SMART_TESTING_MODE) != null) {
+            configuration.mode = RunMode.valueOf(System.getProperty(SMART_TESTING_MODE).toUpperCase());
+        }
+
+        if (System.getProperty(SMART_TESTING_DISABLE) != null) {
+            configuration.disable = Boolean.valueOf(System.getProperty(SMART_TESTING_DISABLE));
+        }
+
+        if (System.getProperty(SMART_TESTING_DEBUG) != null) {
+            configuration.debug = Boolean.valueOf(System.getProperty(SMART_TESTING_DEBUG));
+        }
+
+        configuration.applyTo = System.getProperty(SMART_TESTING_APPLY_TO);
+
+        configuration.report = Report.fromSystemProperties();
+        configuration.scm = Scm.fromSystemProperties();
```

Comment:
To be honest, I still don't like this tightly coupled way of merging the configurations. I just cannot imagine the solution for separated configurations for each strategy. These strategy related configs should be part of the strategy modules and this doesn't allow it. 
I would go for the solution I was proposing you some time ago - having kind of list of objects that contains all necessary information: param name, system property, default value. Then, based on the type (implementing an interface), you could call the method that returns the list of the mentioned objects. This would simplify the logic of getting the system properties (making it generic for all parameters), you could get rid of the `ObjectMapper` class and also simplify this [if](https://github.com/arquillian/smart-testing/pull/183/files#diff-550791242a7d1b9809c8c5ed3e9ac5ecR60). At the same time you could reuse the [convertToType](https://github.com/arquillian/smart-testing/pull/183/files#diff-550791242a7d1b9809c8c5ed3e9ac5ecR94) method for getting the system properties. 
We can have a bjn tomorrow and I can show you what I'm talking about. Some pair programming would make sense as well - I have already some bits implemented.

Reply:
here is the implementation I'm talking about. https://github.com/MatousJobanek/smart-testing/commit/17f94b647ff0feb8b211d8fee9e8d99972d83e62 You can take a look so you know what I have in my mind. It needs some refactoring, removing old implementation classes and methods (such as `fromSystemProperties`) and changes in test classes, but it should work. 
I believe that it will make it simpler and easier to use. 
In addition, I haven't found (in your code) the solution for reading the list of parameters & it's default methods - for Che integration we were talking about. Having the list of configuration items for every configuration section should make it easier to implement.
Let's discuss it in the morning

Reply:
you can find solution for reading the list of parameters in [SchemaGenerator](https://github.com/arquillian/smart-testing/pull/183/files#diff-01b186fdde2ad7c30adc1cfbeb7b08f9)

Reply:
Yeah, I missed that class. Probably because it is package protected and as I need to call it from Che logic I didn't pay much attention to it.
Is there some way of getting default values?

Reply:
You mean default values for each property. for this we need to keep some map with `key` as `property(class's instance field)` & `value` as `default value` for that field.

Reply:
As for the `SchemaGenerator` - maybe it would be good to remove it from this PR and finish it in a separated PR - we can do it together.
This PR is huge enough, so let's finish the main logic first.
I'll create an issue for it